fs/expose/dokan/__init__.py

```diff
@@ -359,111 +404,104 @@ def _check_lock(self, path, offset, length, info, locks=None):
 
     @timeout_protect
     @handle_fs_errors
-    def CreateFile(self, path, access, sharing, disposition, flags, info):
-        path = normpath(path)
+    def ZwCreateFile(self, path, securitycontext, access, attribute, sharing, disposition, options, info):
+        path = self._dokanpath2pyfs(path)
         #  Can't open files that are pending delete.
         if self._is_pending_delete(path):
-            return -ERROR_ACCESS_DENIED
-        # If no access rights are requestsed, only basic metadata is queried.
-        if not access:
-            if self.fs.isdir(path):
-                info.contents.IsDirectory = True
-            elif not self.fs.exists(path):
-                raise ResourceNotFoundError(path)
-            return
-        #  This is where we'd convert the access mask into an appropriate
-        #  mode string.  Unfortunately, I can't seem to work out all the
-        #  details.  I swear MS Word is trying to write to files that it
-        #  opens without asking for write permission.
-        #  For now, just set the mode based on disposition flag.
-        retcode = 0
-        if disposition == CREATE_ALWAYS:
-            if self.fs.exists(path):
-                retcode = ERROR_ALREADY_EXISTS
-            mode = "w+b"
-        elif disposition == OPEN_ALWAYS:
-            if not self.fs.exists(path):
+            return STATUS_ACCESS_DENIED
+
+        retcode = STATUS_SUCCESS
+        if info.contents.IsDirectory:
+            exist = self.fs.exists(path)
+            if disposition == FILE_CREATE:
+                if self.fs.exists(path):
+                    retcode = STATUS_OBJECT_NAME_COLLISION
+                self.fs.makedir(path)
+            elif disposition == FILE_OPEN_IF:
+                if not self.fs.exists(path):
+                    retcode = STATUS_OBJECT_PATH_NOT_FOUND
+        else:
+            # If no access rights are requestsed, only basic metadata is queried.
+            if not access:
+                if self.fs.isdir(path):
+                    info.contents.IsDirectory = True
+                elif not self.fs.exists(path):
+                    return STATUS_OBJECT_NAME_NOT_FOUND
+                return STATUS_SUCCESS
+            #  This is where we'd convert the access mask into an appropriate
+            #  mode string.  Unfortunately, I can't seem to work out all the
+            #  details.  I swear MS Word is trying to write to files that it
+            #  opens without asking for write permission.
+            #  For now, just set the mode based on disposition flag.
+            if disposition == FILE_OVERWRITE_IF or disposition == FILE_SUPERSEDE:
+                if self.fs.exists(path):
+                    retcode = STATUS_OBJECT_NAME_COLLISION
+                mode = "w+b"
+            elif disposition == FILE_OPEN_IF:
+                if not self.fs.exists(path):
+                    mode = "w+b"
+                else:
+                    mode = "r+b"
```

Comment:
The `mode` always contains `+`, meaning it would open file for write. This doesn't work well with read-only file systems :confused: